lib/segment/src/segment_constructor/segment_builder.rs

```diff
@@ -112,117 +125,171 @@ impl SegmentBuilder {
     ///
-    pub fn update_from(&mut self, other: &Segment, stopped: &AtomicBool) -> OperationResult<bool> {
-        self.version = cmp::max(self.version, other.version());
+    pub fn update(
+        &mut self,
+        segments: &[Arc<RwLock<Segment>>],
+        stopped: &AtomicBool,
+    ) -> OperationResult<bool> {
+        if segments.is_empty() {
+            return Ok(true);
+        }
+
+        let segment_guards: Vec<_> = segments.iter().map(|segment| segment.read()).collect();
+
+        let mut merged_points: HashMap<ExtendedPointId, PositionedPointMetadata> = HashMap::new();
+
+        for (segment_index, segment) in segment_guards.iter().enumerate() {
+            for external_id in segment.iter_points() {
+                let version = segment.point_version(external_id).unwrap_or(0);
+                merged_points
+                    .entry(external_id)
+                    .and_modify(|entry| {
+                        if entry.version < version {
+                            entry.segment_index = segment_index;
+                            entry.version = version;
+                        }
+                    })
+                    .or_insert_with(|| {
+                        let internal_id = segment.get_internal_id(external_id).unwrap();
+                        PositionedPointMetadata {
+                            segment_index,
+                            internal_id,
+                            version,
+                        }
+                    });
+            }
+        }
 
-        let other_id_tracker = other.id_tracker.borrow();
-        let other_vector_storages: HashMap<_, _> = other
-            .vector_data
+        let payloads: Vec<_> = segment_guards
             .iter()
-            .map(|(vector_name, vector_data)| {
-                (vector_name.to_owned(), vector_data.vector_storage.borrow())
+            .map(|i| i.payload_index.borrow())
+            .collect();
+
+        let mut points_to_insert: Vec<_> = merged_points
+            .into_iter()
+            .map(|(external_id, pd)| {
+                let mut hasher = AHasher::default();
+
+                // If payload key is set, sort the points to add. For this we fetch the payload and
+                // hash it.
+                let payload_index = self.defragment_key.as_ref().and_then(|defragment_key| {
+                    payloads[pd.segment_index].field_indexes.get(defragment_key)
+                });
+                if let Some(payload_indexes) = payload_index {
+                    // Try to find an index we can use for hashing.
+                    for payload_index in payload_indexes {
+                        match payload_index {
+                            FieldIndex::IntMapIndex(index) => {
+                                if let Some(numbers) = index.get_values(pd.internal_id) {
+                                    for number in numbers {
+                                        number.hash(&mut hasher);
```

Comment:
I have suspicion about the hasher we use, why is it mutable? If hasher is mutable it usually means it accumulates something. For example crc32 blocks in order to get hash for the whole file. But it's opposite of what we need to do here